File: sdks/python/astrodb_client.py
"""
AstroDB Python Client for AI Agents & Astronomical Data Ingestion
"""

import os
import requests
from typing import Dict, Any, List, Optional

class AstroDBClient:
    def __init__(self, base_url: Optional[str] = None, api_key: Optional[str] = None):
        self.base_url = (base_url or os.environ.get("ASTRODB_API_URL") or "http://localhost:3000").rstrip("/")
        self.api_key = api_key or os.environ.get("ASTRODB_API_KEY")
        self.session = requests.Session()
        if self.api_key:
            self.session.headers.update({
                "x-api-key": self.api_key,
                "Authorization": f"Bearer {self.api_key}"
            })

    def search(self, query: str, object_type: Optional[str] = None, constellation: Optional[str] = None, limit: int = 10) -> List[Dict[str, Any]]:
        """
        Hybrid semantic & structured astronomical search
        """
        params = {"q": query, "limit": limit}
        if object_type:
            params["type"] = object_type
        if constellation:
            params["constellation"] = constellation

        resp = self.session.get(f"{self.base_url}/api/v1/search", params=params)
        resp.raise_for_status()
        data = resp.json()
        return data.get("results") or data.get("data") or []

    def get_object(self, object_id_or_name: str) -> Dict[str, Any]:
        """
        Retrieve complete multi-catalog relational dossier for an object
        """
        resp = self.session.get(f"{self.base_url}/api/v1/objects/{object_id_or_name}")
        resp.raise_for_status()
        return resp.json()

    def get_system(self, system_id_or_name: str) -> Dict[str, Any]:
        """
        Retrieve star system architecture (stars + orbiting exoplanets)
        """
        resp = self.session.get(f"{self.base_url}/api/v1/systems/{system_id_or_name}")
        resp.raise_for_status()
        return resp.json()

    def query_catalog(self, catalog_type: str, in_habitable_zone: bool = False, min_esi: Optional[float] = None, limit: int = 50) -> List[Dict[str, Any]]:
        """
        Query specialized sub-catalogs (exoplanets, stars, astrobiology, galaxies, relics)
        """
        params = {"limit": limit}
        if in_habitable_zone:
            params["in_habitable_zone"] = "true"
        if min_esi is not None:
            params["min_esi"] = min_esi

        resp = self.session.get(f"{self.base_url}/api/v1/catalogs/{catalog_type}", params=params)
        resp.raise_for_status()
        return resp.json().get("data", [])

    def query_sql(self, select_query: str) -> List[Dict[str, Any]]:
        """
        Execute safe read-only SQL query against the AstroDB sandbox
        """
        resp = self.session.post(f"{self.base_url}/api/v1/query", json={"query": select_query})
        resp.raise_for_status()
        return resp.json().get("data", [])

    def get_tools_schema(self) -> Dict[str, Any]:
        """
        Get JSON tool definitions for OpenAI Function Calling & LangChain
        """
        resp = self.session.get(f"{self.base_url}/api/v1/tools")
        resp.raise_for_status()
        return resp.json()

File: sdks/typescript/astrodb-client.ts
/**
 * AstroDB TypeScript Client for AI Agents and Next.js / Node runtimes
 */

export interface AstroDBConfig {
  baseUrl?: string;
  apiKey?: string;
}

export class AstroDBClient {
  private baseUrl: string;
  private apiKey?: string;

  constructor(config: AstroDBConfig = {}) {
    this.baseUrl = (config.baseUrl || process.env.ASTRODB_API_URL || 'http://localhost:3000').replace(/\/$/, '');
    this.apiKey = config.apiKey || process.env.ASTRODB_API_KEY;
  }

  private async request<T>(path: string, options: RequestInit = {}): Promise<T> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      ...(options.headers as Record<string, string> || {}),
    };

    if (this.apiKey) {
      headers['x-api-key'] = this.apiKey;
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    const res = await fetch(`${this.baseUrl}${path}`, {
      ...options,
      headers,
    });

    if (!res.ok) {
      const err = await res.json().catch(() => ({ error: res.statusText }));
      throw new Error(err.error || `AstroDB request failed with status ${res.status}`);
    }

    return res.json();
  }

  /**
   * Hybrid semantic & structured search
   */
  async search(query: string, options: { objectType?: string; constellation?: string; limit?: number } = {}) {
    const params = new URLSearchParams({ q: query, limit: String(options.limit || 10) });
    if (options.objectType) params.append('type', options.objectType);
    if (options.constellation) params.append('constellation', options.constellation);

    const res: any = await this.request(`/api/v1/search?${params.toString()}`);
    return res.results || res.data || [];
  }

  /**
   * Get complete object dossier
   */
  async getObject(idOrName: string) {
    return this.request(`/api/v1/objects/${encodeURIComponent(idOrName)}`);
  }

  /**
   * Get star system architecture
   */
  async getSystem(systemIdOrName: string) {
    return this.request(`/api/v1/systems/${encodeURIComponent(systemIdOrName)}`);
  }

  /**
   * Safe read-only SQL query execution
   */
  async querySql(query: string) {
    const res: any = await this.request('/api/v1/query', {
      method: 'POST',
      body: JSON.stringify({ query }),
    });
    return res.data || [];
  }
}
